File: sol/solution/0001-0100/0037. Sudoku Solver/Solution.scala
import scala.collection.mutable.ArrayBuffer
object Solution {
    def solveSudoku(board: Array[Array[Char]]): Unit = {
        val rows = Array.fill(9)(0)
        val cols = Array.fill(9)(0)
        val boxes = Array.fill(9)(0)
        val empties = ArrayBuffer.empty[Int]
        var r = 0
        while (r < 9) {
            var c = 0
            while (c < 9) {
                val ch = board(r)(c)
                if (ch == '.') {
                    empties += (r * 9 + c)
                } else {
                    val d = ch - '1'
                        val bit = 1 << d
                    rows(r) |= bit
                    cols(c) |= bit
                    boxes((r / 3) * 3 + (c / 3)) |= bit
                }
                c += 1
            }
            r += 1
        }
        def dfs(): Boolean = {
            if (empties.isEmpty) return true
            var bestIdx = -1
            var bestMask = 0
            var bestCount = 10
            var i = 0
            while (i < empties.length) {
                val pos = empties(i)
                val rr = pos / 9
                val cc = pos % 9
                val bb = (rr / 3) * 3 + (cc / 3)
                val used = rows(rr) | cols(cc) | boxes(bb)
                val mask = (~used) & 0x1FF
                val cnt = java.lang.Integer.bitCount(mask)
                if (cnt == 0) return false
                if (cnt < bestCount) {
                    bestCount = cnt
                    bestMask = mask
                    bestIdx = i
                    if (cnt == 1) {}
                }
                i += 1
            }
            val last = empties.length - 1
            val chosen = empties(bestIdx)
            empties(bestIdx) = empties(last)
            empties.remove(last)
            val rr = chosen / 9
            val cc = chosen % 9
            val bb = (rr / 3) * 3 + (cc / 3)
            var mask = bestMask
            while (mask != 0) {
                val bit = mask & -mask
                val d = java.lang.Integer.numberOfTrailingZeros(bit)
                board(rr)(cc) = (d + '1').toChar
                rows(rr) |= bit
                cols(cc) |= bit
                boxes(bb) |= bit
                if (dfs()) return true
                rows(rr) &= ~bit
                cols(cc) &= ~bit
                boxes(bb) &= ~bit
                mask &= (mask - 1)
            }
            board(rr)(cc) = '.'
            empties += chosen
            false
        }
        dfs()
    }
}

File: sol/solution/0001-0100/0037. Sudoku Solver/Solution.py
class Solution:
    def solveSudoku(self, board: list[list[str]]) -> None:
        rows = [0] * 9
        cols = [0] * 9
        boxes = [0] * 9
        empties = []
        for r in range(9):
            for c in range(9):
                ch = board[r][c]
                if ch == '.':
                    empties.append((r, c))
                else:
                    bit = 1 << (ord(ch) - ord('1'))
                    rows[r] |= bit
                    cols[c] |= bit
                    boxes[(r // 3) * 3 + c // 3] |= bit
        def backtrack(idx=0):
            if idx == len(empties):
                return True
            best = idx
            min_options = 10
            for k in range(idx, len(empties)):
                r, c = empties[k]
                b = (r // 3) * 3 + c // 3
                mask = ~(rows[r] | cols[c] | boxes[b]) & 0x1FF
                options = mask.bit_count()
                if options < min_options:
                    min_options = options
                    best = k
                    if options == 1:
                        break
            if min_options == 0:
                return False
            empties[idx], empties[best] = empties[best], empties[idx]
            r, c = empties[idx]
            b = (r // 3) * 3 + c // 3
            mask = ~(rows[r] | cols[c] | boxes[b]) & 0x1FF
            while mask:
                bit = mask & -mask
                mask -= bit
                d = bit.bit_length()
                ch = str(d)
                board[r][c] = ch
                rows[r] |= bit
                cols[c] |= bit
                boxes[b] |= bit
                if backtrack(idx + 1):
                    return True
                board[r][c] = '.'
                rows[r] &= ~bit
                cols[c] &= ~bit
                boxes[b] &= ~bit
            return False
        backtrack()